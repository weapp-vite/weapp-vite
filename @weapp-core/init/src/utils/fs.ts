import fs from 'fs-extra'

export async function readJsonIfExists<T>(filepath: string): Promise<T | null> {
  try {
    if (!await fs.pathExists(filepath)) {
      return null
    }
    return await fs.readJSON(filepath) as T
  }
  catch (error) {
    throw new FsReadError(filepath, error)
  }
}

export async function readFileIfExists(filepath: string): Promise<string | null> {
  try {
    if (!await fs.pathExists(filepath)) {
      return null
    }
    return await fs.readFile(filepath, 'utf8')
  }
  catch (error) {
    throw new FsReadError(filepath, error)
  }
}

export async function writeJsonFile(filepath: string, data: unknown, spaces = 2) {
  try {
    await fs.outputJSON(filepath, data, {
      spaces,
    })
  }
  catch (error) {
    throw new FsWriteError(filepath, error)
  }
}

export async function writeFile(filepath: string, contents: string) {
  try {
    await fs.outputFile(filepath, contents, 'utf8')
  }
  catch (error) {
    throw new FsWriteError(filepath, error)
  }
}

export class FsReadError extends Error {
  constructor(public filepath: string, public cause: unknown) {
    super(`Failed to read ${filepath}`)
    this.name = 'FsReadError'
  }
}

export class FsWriteError extends Error {
  constructor(public filepath: string, public cause: unknown) {
    super(`Failed to write ${filepath}`)
    this.name = 'FsWriteError'
  }
}
