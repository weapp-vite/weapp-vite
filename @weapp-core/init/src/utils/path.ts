import path from 'pathe'

export function resolveOutputPath(root: string, dest: string | undefined, fallback: string) {
  if (!dest) {
    return fallback
  }

  return path.isAbsolute(dest) ? dest : path.resolve(root, dest)
}
