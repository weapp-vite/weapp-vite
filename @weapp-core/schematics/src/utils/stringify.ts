export function toJsonString(value: unknown) {
  return JSON.stringify(value, undefined, 2)
}
