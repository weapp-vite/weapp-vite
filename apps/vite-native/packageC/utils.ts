export function getPackageName() {
  return 'xxxx'
}
