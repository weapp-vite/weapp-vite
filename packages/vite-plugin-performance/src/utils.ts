export function ensureArray<T>(value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value]
}

export function isPromiseLike<T>(value: T | Promise<T>): value is Promise<T> {
  return typeof value === 'object' && value !== null && 'then' in value
}
