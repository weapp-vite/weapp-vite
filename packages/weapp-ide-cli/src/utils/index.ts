export * from './argv'
export * from './exec'
export * from './path'
