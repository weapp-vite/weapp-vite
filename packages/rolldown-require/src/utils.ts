import type { DepOptimizationOptions } from './optimizer'
import type { PackageCache } from './packages'
import type { RequireFunction } from './types'
import { exec } from 'node:child_process'
import fs from 'node:fs'
import { builtinModules, createRequire } from 'node:module'
import path from 'node:path'
import process from 'node:process'
import { createFilter as _createFilter } from '@rollup/pluginutils'
import {
  OPTIMIZABLE_ENTRY_RE,
} from './constants'
import { findNearestPackageData } from './packages'
import {
  isWindows,
  slash,
  splitFileAndPostfix,
} from './sharedUtils'

/**
 * Inlined to keep `@rollup/pluginutils` in devDependencies
 */
export type FilterPattern
  = | ReadonlyArray<string | RegExp>
    | string
    | RegExp
    | null
export const createFilter = _createFilter as (
  include?: FilterPattern,
  exclude?: FilterPattern,
  options?: { resolve?: string | false | null },
) => (id: string | unknown) => boolean

// Supported by Node, Deno, Bun
const NODE_BUILTIN_NAMESPACE = 'node:'
// Supported by Deno
const NPM_BUILTIN_NAMESPACE = 'npm:'
// Supported by Bun
const BUN_BUILTIN_NAMESPACE = 'bun:'
// Some runtimes like Bun injects namespaced modules here, which is not a node builtin
const nodeBuiltins = builtinModules.filter(id => !id.includes(':'))

const isBuiltinCache = new WeakMap<
  (string | RegExp)[],
  (id: string, importer?: string) => boolean
>()

export function isBuiltin(builtins: (string | RegExp)[], id: string): boolean {
  let isBuiltin = isBuiltinCache.get(builtins)
  if (!isBuiltin) {
    isBuiltin = createIsBuiltin(builtins)
    isBuiltinCache.set(builtins, isBuiltin)
  }
  return isBuiltin(id)
}

export function createIsBuiltin(
  builtins: (string | RegExp)[],
): (id: string) => boolean {
  const plainBuiltinsSet = new Set(
    builtins.filter(builtin => typeof builtin === 'string'),
  )
  const regexBuiltins = builtins.filter(
    builtin => typeof builtin !== 'string',
  )

  return id =>
    plainBuiltinsSet.has(id) || regexBuiltins.some(regexp => regexp.test(id))
}

export const nodeLikeBuiltins = [
  ...nodeBuiltins,
  new RegExp(`^${NODE_BUILTIN_NAMESPACE}`),
  new RegExp(`^${NPM_BUILTIN_NAMESPACE}`),
  new RegExp(`^${BUN_BUILTIN_NAMESPACE}`),
]

export function isNodeLikeBuiltin(id: string): boolean {
  return isBuiltin(nodeLikeBuiltins, id)
}

export function isNodeBuiltin(id: string): boolean {
  if (id.startsWith(NODE_BUILTIN_NAMESPACE)) {
    return true
  }
  return nodeBuiltins.includes(id)
}

export function isInNodeModules(id: string): boolean {
  return id.includes('node_modules')
}

export function isOptimizable(
  id: string,
  optimizeDeps: DepOptimizationOptions,
): boolean {
  const { extensions } = optimizeDeps
  return (
    OPTIMIZABLE_ENTRY_RE.test(id)
    || (extensions?.some(ext => id.endsWith(ext)) ?? false)
  )
}

export const bareImportRE = /^(?![a-z]:)[\w@](?!.*:\/\/)/i
export const deepImportRE = /^([^@][^/]*)\/|^(@[^/]+\/[^/]+)\//

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

export function injectQuery(url: string, queryToInject: string): string {
  const { file, postfix } = splitFileAndPostfix(url)
  const normalizedFile = isWindows ? slash(file) : file
  return `${normalizedFile}?${queryToInject}${postfix[0] === '?' ? `&${postfix.slice(1)}` : /* hash only */ postfix}`
}

export function isObject(value: unknown): value is Record<string, any> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export function tryStatSync(file: string): fs.Stats | undefined {
  try {
    // The "throwIfNoEntry" is a performance optimization for cases where the file does not exist
    return fs.statSync(file, { throwIfNoEntry: false })
  }
  catch {
    // Ignore errors
  }
}

export function isFilePathESM(
  filePath: string,
  packageCache?: PackageCache,
): boolean {
  if (/\.m[jt]s$/.test(filePath)) {
    return true
  }
  else if (/\.c[jt]s$/.test(filePath)) {
    return false
  }
  else {
    // check package.json for type: "module"
    try {
      const pkg = findNearestPackageData(path.dirname(filePath), packageCache)
      return pkg?.data.type === 'module'
    }
    catch {
      return false
    }
  }
}

// `fs.realpathSync.native` resolves differently in Windows network drive,
// causing file read errors. skip for now.
// https://github.com/nodejs/node/issues/37737
let currentSafeRealpathSync = isWindows
  ? windowsSafeRealPathSync
  : fs.realpathSync.native

export function safeRealpathSync(filePath: string): string {
  return currentSafeRealpathSync(filePath)
}

// Based on https://github.com/larrybahr/windows-network-drive
// MIT License, Copyright (c) 2017 Larry Bahr
const windowsNetworkMap = new Map()
function windowsMappedRealpathSync(path: string) {
  const realPath = fs.realpathSync.native(path)
  if (realPath.startsWith('\\\\')) {
    for (const [network, volume] of windowsNetworkMap) {
      if (realPath.startsWith(network)) { return realPath.replace(network, volume) }
    }
  }
  return realPath
}
const parseNetUseRE = /^\w* +(\w:) +([^ ]+)\s/
let firstSafeRealPathSyncRun = false

function windowsSafeRealPathSync(path: string): string {
  if (!firstSafeRealPathSyncRun) {
    optimizeSafeRealPathSync()
    firstSafeRealPathSyncRun = true
  }
  return fs.realpathSync(path)
}

function optimizeSafeRealPathSync() {
  // Skip if using Node <18.10 due to MAX_PATH issue: https://github.com/vitejs/vite/issues/12931
  const nodeVersion = process.versions.node.split('.').map(Number)
  if (nodeVersion[0] < 18 || (nodeVersion[0] === 18 && nodeVersion[1] < 10)) {
    currentSafeRealpathSync = fs.realpathSync
    return
  }
  // Check the availability `fs.realpathSync.native`
  // in Windows virtual and RAM disks that bypass the Volume Mount Manager, in programs such as imDisk
  // get the error EISDIR: illegal operation on a directory
  try {
    fs.realpathSync.native(path.resolve('./'))
  }
  catch (error) {
    // @ts-ignore
    if (error.message.includes('EISDIR: illegal operation on a directory')) {
      currentSafeRealpathSync = fs.realpathSync
      return
    }
  }
  exec('net use', (error, stdout) => {
    if (error) { return }
    const lines = stdout.split('\n')
    // OK           Y:        \\NETWORKA\Foo         Microsoft Windows Network
    // OK           Z:        \\NETWORKA\Bar         Microsoft Windows Network
    for (const line of lines) {
      const m = parseNetUseRE.exec(line)
      if (m) { windowsNetworkMap.set(m[2], m[1]) }
    }
    currentSafeRealpathSync
      = windowsNetworkMap.size === 0
        ? fs.realpathSync.native
        : windowsMappedRealpathSync
  })
}

// strip UTF-8 BOM
export function stripBomTag(content: string): string {
  if (content.charCodeAt(0) === 0xFEFF) {
    return content.slice(1)
  }

  return content
}

export function getNpmPackageName(importPath: string): string | null {
  const parts = importPath.split('/')
  if (parts[0][0] === '@') {
    if (!parts[1]) { return null }
    return `${parts[0]}/${parts[1]}`
  }
  else {
    return parts[0]
  }
}

// Injected by TSUP
declare const TSUP_FORMAT: 'esm' | 'cjs'
export const dynamicImport: RequireFunction = async (
  id: string,
  { format },
) => {
  const fn
    = format === 'esm'
      ? (file: string) => import(file)
      : TSUP_FORMAT === 'esm'
        ? createRequire(import.meta.url)
        : require
  return fn(id)
}
