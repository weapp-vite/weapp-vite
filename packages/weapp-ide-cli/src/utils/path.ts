import process from 'node:process'
import path from 'pathe'

export function resolvePath(filePath: string) {
  if (path.isAbsolute(filePath)) {
    return filePath
  }

  return path.resolve(process.cwd(), filePath)
}
