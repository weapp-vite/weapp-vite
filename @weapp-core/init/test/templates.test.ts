import path from 'node:path'
import { scanFiles } from './utils'

describe('templates', () => {
  it('should ', async () => {
    await (await import('../scripts/shared')).main()
    const files = await scanFiles(path.resolve(__dirname, '../templates'))
    expect(files).toMatchSnapshot()
  })
})
