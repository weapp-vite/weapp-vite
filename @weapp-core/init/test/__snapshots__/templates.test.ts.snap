// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`templates > should  1`] = `
[
  "default/.editorconfig",
  "default/.gitignore",
  "default/.vscode/settings.json",
  "default/package.json",
  "default/project.config.json",
  "default/project.private.config.json",
  "default/public/logo.png",
  "default/README.md",
  "default/src/app.json",
  "default/src/app.scss",
  "default/src/app.ts",
  "default/src/components/Navbar/Navbar.json",
  "default/src/components/Navbar/Navbar.scss",
  "default/src/components/Navbar/Navbar.ts",
  "default/src/components/Navbar/Navbar.wxml",
  "default/src/pages/index/index.scss",
  "default/src/pages/index/index.ts",
  "default/src/pages/index/index.wxml",
  "default/src/sitemap.json",
  "default/src/theme.json",
  "default/src/utils/util.ts",
  "default/src/vite-env.d.ts",
  "default/tsconfig.json",
  "default/tsconfig.node.json",
  "default/vite.config.ts",
  "tailwindcss/.editorconfig",
  "tailwindcss/.gitignore",
  "tailwindcss/.vscode/settings.json",
  "tailwindcss/package.json",
  "tailwindcss/postcss.config.js",
  "tailwindcss/project.config.json",
  "tailwindcss/project.private.config.json",
  "tailwindcss/public/icon0.png",
  "tailwindcss/public/icon0s.png",
  "tailwindcss/public/icon1.png",
  "tailwindcss/public/icon1s.png",
  "tailwindcss/public/logo.png",
  "tailwindcss/README.md",
  "tailwindcss/src/app.json",
  "tailwindcss/src/app.scss",
  "tailwindcss/src/app.ts",
  "tailwindcss/src/components/Navbar/Navbar.json",
  "tailwindcss/src/components/Navbar/Navbar.scss",
  "tailwindcss/src/components/Navbar/Navbar.ts",
  "tailwindcss/src/components/Navbar/Navbar.wxml",
  "tailwindcss/src/pages/index/index.json",
  "tailwindcss/src/pages/index/index.scss",
  "tailwindcss/src/pages/index/index.ts",
  "tailwindcss/src/pages/index/index.wxml",
  "tailwindcss/src/pages/profile/index.scss",
  "tailwindcss/src/pages/profile/index.ts",
  "tailwindcss/src/pages/profile/index.wxml",
  "tailwindcss/src/sitemap.json",
  "tailwindcss/src/theme.json",
  "tailwindcss/src/utils/util.ts",
  "tailwindcss/src/vite-env.d.ts",
  "tailwindcss/tailwind.config.ts",
  "tailwindcss/tsconfig.json",
  "tailwindcss/tsconfig.node.json",
  "tailwindcss/vite.config.ts",
  "tdesign/.editorconfig",
  "tdesign/.gitignore",
  "tdesign/.vscode/settings.json",
  "tdesign/package.json",
  "tdesign/postcss.config.js",
  "tdesign/project.config.json",
  "tdesign/project.private.config.json",
  "tdesign/public/logo.png",
  "tdesign/README.md",
  "tdesign/src/app.json",
  "tdesign/src/app.scss",
  "tdesign/src/app.ts",
  "tdesign/src/components/Navbar/Navbar.json",
  "tdesign/src/components/Navbar/Navbar.scss",
  "tdesign/src/components/Navbar/Navbar.ts",
  "tdesign/src/components/Navbar/Navbar.wxml",
  "tdesign/src/pages/index/index.json",
  "tdesign/src/pages/index/index.scss",
  "tdesign/src/pages/index/index.ts",
  "tdesign/src/pages/index/index.wxml",
  "tdesign/src/sitemap.json",
  "tdesign/src/theme.json",
  "tdesign/src/utils/util.ts",
  "tdesign/src/vite-env.d.ts",
  "tdesign/tailwind.config.ts",
  "tdesign/tsconfig.json",
  "tdesign/tsconfig.node.json",
  "tdesign/vite.config.ts",
  "vant/.editorconfig",
  "vant/.gitignore",
  "vant/.vscode/settings.json",
  "vant/package.json",
  "vant/postcss.config.js",
  "vant/project.config.json",
  "vant/project.private.config.json",
  "vant/public/logo.png",
  "vant/README.md",
  "vant/src/app.json",
  "vant/src/app.scss",
  "vant/src/app.ts",
  "vant/src/components/Navbar/Navbar.json",
  "vant/src/components/Navbar/Navbar.scss",
  "vant/src/components/Navbar/Navbar.ts",
  "vant/src/components/Navbar/Navbar.wxml",
  "vant/src/pages/index/index.json",
  "vant/src/pages/index/index.scss",
  "vant/src/pages/index/index.ts",
  "vant/src/pages/index/index.wxml",
  "vant/src/sitemap.json",
  "vant/src/theme.json",
  "vant/src/utils/util.ts",
  "vant/src/vite-env.d.ts",
  "vant/tailwind.config.ts",
  "vant/tsconfig.json",
  "vant/tsconfig.node.json",
  "vant/vite.config.ts",
]
`;
